### ISSUE TITLE
Incremental parsing issue with namespace packages

### ISSUE THREAD

**Issue by PeterJCLaw** [python/mypy#12664]
Given a local file which uses something which is a namespace package and then another local file which uses the first one, there's an intermittent bug where the namespace packaged module's types aren't found.
This seems likely similar to https://github.com/python/mypy/issues/7276, however even with https://github.com/python/mypy/pull/10937 in place (i.e: on a more recent mypy release) the issue remains.

**To Reproduce**

``` shell
$ mkdir demo
$ echo 'import ruamel.yaml' > demo/wrapper.py 
$ echo 'import wrapper' > demo/consumer.py
```
``` shell
$ mypy demo  # first run
Success: no issues found in 2 source files
```
``` shell
$ mypy demo  # second run
demo/wrapper.py:1: error: Skipping analyzing "ruamel": module is installed, but missing library stubs or py.typed marker
demo/consumer.py:1: error: Skipping analyzing "ruamel": module is installed, but missing library stubs or py.typed marker
demo/consumer.py:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
Found 2 errors in 2 files (checked 2 source files)
```

Subsequent runs will alternate between success and failure. Any additional source files in the project (whether they import the `wrapper` or not) will also have the error message applied to them, which makes failing runs vastly less useful.

Running with `--no-incremental` appears to avoid the issue; I can run several times in a row with `--no-incremental` without encountering the failure mode (and have not seen it with that flag enabled).

Running with `--namespace-packages` appears to have no effect.

Unfortunately as `ruamel.yaml` is a typed library, ignoring the imports with `type: ignore[import]` ends up generating more errors as the ignores appear unused on a "successful" run (and you have `warn_unused_ignores` enabled).

<details>
<summary>Setup & versions</summary>

``` shell
python3.9 -m venv venv
. venv/bin/activate
pip install -U pip setuptools wheel
pip install mypy ruamel.yaml
```
``` shell
$ pip list
Package           Version
----------------- -------
mypy              0.942
mypy-extensions   0.4.3
pip               22.0.4
ruamel.yaml       0.17.21
ruamel.yaml.clib  0.2.6
setuptools        62.1.0
tomli             2.0.1
typing_extensions 4.2.0
wheel             0.37.1
```

</details>

**Your Environment**

- Mypy version used: 0.942
- Mypy command-line flags: see above
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.9 (though similar behaviour observed on other versions; at least on 3.8)
- Operating system and version: Ubuntu 20.04 LTS

**Comment by hauntsaninja** [comment]
Thanks, I can repro on master, which means the unreleased https://github.com/python/mypy/pull/12250 does not fix either

**Comment by jacleggmsr** [comment]
I'm getting the same issue.

I'm running python 3.10.7, mypy 0.982, mypy-extensions 0.4.3, ruamel.yaml 0.17.21 and ruamel.yaml.clib 0.2.7

**Comment by davnat** [comment]
I can confirm this on Ubuntu 20, python 3.8.10, mypy 1.3.0.

In my case the problem appears only in a package that depends on ruamel-yaml very indirectly (it's a dep of a dep of a dep), while in the original package that has the direct dependency and all others in the chain all seems OK, for now.

I noted that the problematic package has a lot of `from foo import *`, maybe this could trigger the error?

**Comment by dahlbaek** [comment]
I can confirm this issue on MacOS 13.2.1, python 3.10.10, mypy 1.8.0

**Comment by DBCerigo** [comment]
Similarly still seeing this issue on intermitant `mypy` runs. `--no-incremental` fixes it, but also seems to slow down the typechecking significantly (as I think is to be expected), so is far from an ideal solution. 

On `Ubuntu Debian 22.04 LTS`
```
mypy==1.8.0
ruamel.yaml==0.18.6
ruamel.yaml.clib==0.2.8
```

**Comment by Sh4pe** [comment]
We are affected by this issue as well. Using `--no-incremental` or simply deleting the `.mypy_cache` directory each time works as a workaround, but this slows down type checking considerably. Which is obviously not preferred.

We are using Poetry and this are the relevant versions from our poetry.lock:

```
mypy              1.10.0
ruamel-yaml       0.18.6
ruamel-yaml-clib  0.2.8
```

I want to say that using mypy is very valuable to us, but this is a small annoyance. Is there a workaround that works with the cache?

**Comment by petergaultney** [comment]
```
python 3.10.13
mypy 1.10.1
google-auth 2.30.0
```

Likely not relevant, but this does not ever appear to happen with `.pth` local editable dependencies that are themselves namespace packages. We have lots of those and this issue never occurs with them.

It's also not deterministic, but I can confirm the behavior where it always succeeds when the cache is empty.

**Comment by rworkduo** [comment]
after a year, this issue is still haunting me and costing me a significant amount of CI running time
is it even being looked into yet?

**Comment by martinhoyer** [comment]
This is ugly, but avoids usage of `--no-incremental`, and only skips the cache for problematic modules.
In our case it's ruamel.yaml - `mypy.ini`: 
```
[ruamel.*]
cache_dir = /dev/null
```
(the ruamel/yaml cache is still created, but not used in subsequent runs, IIUC)

**Comment by ohshazbot** [comment]
> This is ugly, but avoids usage of `--no-incremental`, and only skips the cache for problematic modules. In our case it's ruamel.yaml - `mypy.ini`:
> 
> ```
> [ruamel.*]
> cache_dir = /dev/null
> ```
> 
> (the ruamel/yaml cache is still created, but not used in subsequent runs, IIUC)

unfortunately this solution does not seem to work for me, it does not cause any sort of behavior change

**Comment by ego-thales** [comment]
Thank god for the OP perfectly describing the issue. I encounter it here.

Using `uv 0.6.6`, `mypy v1.15.0`, `ruamel-yaml v0.18.10` on Ubuntu 22.04 and python 3.12/3.13.

All the best.

**Comment by hauntsaninja** [comment]
One hack "fix":
```
[[tool.mypy.overrides]]
module = "ruamel.yaml"
follow_imports = "skip"
```

**Comment by ego-thales** [comment]
Thanks @hauntsaninja! This alternative monkeypatch is much more appropriate to me. It allows caching to work properly only at the cost of not following this specific package, which does not really need type checking in my project.

It'll save me a lot of time <3

**Comment by hauntsaninja** [comment]
This is released now in 1.16
